With net6.0 preview 3, publish console app with RID(osx-x64) failed with error MSB4018.

Repro steps:
1. Install SDK 6.0 preview 3 from master branch of https://github.com/dotnet/installer
2. mkdir hello
3. dotnet new console
4. dotnet publish -c Release -r osx-x64

Expected Result:
Publish with RID(osx-x64) should work fine.

Actual Result:
Publish with RID(osx-x64) failed with error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [C:\Users\v-damu\hello1\hello1.csproj]
![image](https://user-images.githubusercontent.com/65638819/110762384-b4c7f380-828b-11eb-939a-a5533a44d0e3.png)

Note:
Publish with RID(win-x64, linux-x64) is working fine.

dotnet --info:
.NET SDK (reflecting any global.json):
Version: 6.0.100-preview.3.21160.18
Commit: 5a0f4f6800

Runtime Environment:
OS Name: Windows
OS Version: 10.0.19041
OS Platform: Windows
RID: win10-x64
Base Path: C:\Program Files\dotnet\sdk\6.0.100-preview.3.21160.18\

Host (useful for support):
Version: 6.0.0-preview.3.21159.16
Commit: 3eddf4a06b

.NET SDKs installed:
6.0.100-preview.3.21160.18 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
Microsoft.AspNetCore.App 6.0.0-preview.3.21160.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.NETCore.App 6.0.0-preview.3.21159.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.WindowsDesktop.App 6.0.0-preview.3.21127.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
https://aka.ms/dotnet-download

Ah, I was testing the wrong binary when copied to macOS, here is what happens when I run that app (built after commenting out aforementioned two lines):

```sh
$ ./hello
dyld: malformed mach-o image: symbol strings overrun __LINKEDIT
bort trap: 6           ./hello
Error: Process completed with exit code 134.
```

So at least our verification is legit.

--- a/src/installer/managed/Microsoft.NET.HostModel/AppHost/BinaryUtils.cs
+++ b/src/installer/managed/Microsoft.NET.HostModel/AppHost/BinaryUtils.cs
@@ -174,15 +174,16 @@ namespace Microsoft.NET.HostModel.AppHost
             File.Copy(sourcePath, destinationPath, overwrite: true);
         }
 
-        internal static void WriteToStream(MemoryMappedViewAccessor sourceViewAccessor, FileStream fileStream)
+        internal static void WriteToStream(MemoryMappedViewAccessor sourceViewAccessor, FileStream fileStream, long length)
         {
             int pos = 0;
             int bufSize = 16384; //16K
 
             byte[] buf = new byte[bufSize];
+            length = Math.Min(length, sourceViewAccessor.Capacity);
             do
             {
-                int bytesRequested = Math.Min((int)sourceViewAccessor.Capacity - pos, bufSize);
+                int bytesRequested = Math.Min((int)length - pos, bufSize);
                 if (bytesRequested <= 0)
                 {
                     break;

--- a/src/installer/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs
+++ b/src/installer/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs
@@ -85,13 +85,19 @@ namespace Microsoft.NET.HostModel.AppHost
             {
                 RetryUtil.RetryOnIOError(() =>
                 {
+                    FileStream appHostSourceStream = null;
                     MemoryMappedFile memoryMappedFile = null;
                     MemoryMappedViewAccessor memoryMappedViewAccessor = null;
                     try
                     {
                         // Open the source host file.
-                        memoryMappedFile = MemoryMappedFile.CreateFromFile(appHostSourceFilePath, FileMode.Open, null, 0, MemoryMappedFileAccess.Read);
+                        appHostSourceStream = new FileStream(appHostSourceFilePath, FileMode.Open, FileAccess.Read, FileShare.Read);
+                        memoryMappedFile = MemoryMappedFile.CreateFromFile(appHostSourceStream, null, 0, MemoryMappedFileAccess.Read, HandleInheritability.None, true);
                         memoryMappedViewAccessor = memoryMappedFile.CreateViewAccessor(0, 0, MemoryMappedFileAccess.CopyOnWrite);
+
+                        // Get the size of the source app host to ensure that we don't write extra data to the destination.
+                        // On Windows, the size of the view accessor is rounded up to the next page boundary.
+                        long sourceAppHostLength = appHostSourceStream.Length;
 
                         // Transform the host file in-memory.
                         RewriteAppHost(memoryMappedViewAccessor);
@@ -99,7 +105,7 @@ namespace Microsoft.NET.HostModel.AppHost
                         // Save the transformed host.
                         using (FileStream fileStream = new FileStream(appHostDestinationFilePath, FileMode.Create))
                         {
-                            BinaryUtils.WriteToStream(memoryMappedViewAccessor, fileStream);
+                            BinaryUtils.WriteToStream(memoryMappedViewAccessor, fileStream, sourceAppHostLength);
 
                             // Remove the signature from MachO hosts.
                             if (!appHostIsPEImage)
@@ -112,6 +118,7 @@ namespace Microsoft.NET.HostModel.AppHost
                     {
                         memoryMappedViewAccessor?.Dispose();
                         memoryMappedFile?.Dispose();
+                        appHostSourceStream?.Dispose();
                     }
                 });
 

